source/extensions/transport_sockets/http_11_proxy/connect.cc

```diff
@@ -0,0 +1,140 @@
+  if (header_buffer_.length() > 0) {
+    auto header_res = writeHeader();
+    return header_res;
+  } else if (stripped_connect_) {
+    // Don't pass events up until the connect response is read because TLS reads
+    // kick off writes which don't pass through the transport socket.
+    return transport_socket_->doWrite(buffer, end_stream);
+  }
+  return Network::IoResult{Network::PostIoAction::KeepOpen, 0, false};
+}
+
+Network::IoResult UpstreamHttp11ConnectSocket::doRead(Buffer::Instance& buffer) {
+  if (!stripped_connect_) {
+    char peek_buf[200];
+    auto result = callbacks_->ioHandle().recv(peek_buf, 200, MSG_PEEK);
+    if (!result.ok() && result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+      return {Network::PostIoAction::Close, 0, false};
+    }
+    absl::string_view peek_data(peek_buf, result.return_value_);
+    size_t index = peek_data.find("\r\n\r\n");
+    if (index == absl::string_view::npos) {
+      if (result.return_value_ == 200) {
+        ENVOY_CONN_LOG(trace, "failed to receive CONNECT headers within 200 bytes",
+                       callbacks_->connection());
+        return {Network::PostIoAction::Close, 0, false};
+      }
+      return Network::IoResult{Network::PostIoAction::KeepOpen, 0, false};
+    } else {
+      result = callbacks_->ioHandle().read(buffer, index + 4);
+      if (!result.ok() || result.return_value_ != index + 4) {
+        ENVOY_CONN_LOG(trace, "failed to drain CONNECT header", callbacks_->connection());
+        return {Network::PostIoAction::Close, 0, false};
+      }
+      // Note this is not in any way proper HTTP/1.1 parsing.
+      // Before this is used with any untrusted upstream, proper checks should
+      // be done rather than this.
+      if (!absl::StartsWith(peek_data, "HTTP/1.1 200")) {
+        ENVOY_CONN_LOG(trace, "Response does not match strict connect checks",
+                       callbacks_->connection());
+        return {Network::PostIoAction::Close, 0, false};
+      }
+      buffer.drain(buffer.length());
+      stripped_connect_ = true;
+    }
+  }
+  return transport_socket_->doRead(buffer);
+}
+
+Network::IoResult UpstreamHttp11ConnectSocket::writeHeader() {
+  Network::PostIoAction action = Network::PostIoAction::KeepOpen;
+  uint64_t bytes_written = 0;
+  do {
+    if (header_buffer_.length() == 0) {
+      break;
+    }
+
+    Api::IoCallUint64Result result = callbacks_->ioHandle().write(header_buffer_);
+
+    if (result.ok()) {
+      ENVOY_CONN_LOG(trace, "write returns: {}", callbacks_->connection(), result.return_value_);
+      bytes_written += result.return_value_;
+    } else {
+      ENVOY_CONN_LOG(trace, "write error: {}", callbacks_->connection(),
+                     result.err_->getErrorDetails());
+      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+        action = Network::PostIoAction::Close;
+      }
+      break;
```

Comment:
It'd be inclined to avoid a bit of nesting and write this as:
```
if (!result.ok()) {
      ENVOY_CONN_LOG(trace, "write error: {}", callbacks_->connection(),
                     result.err_->getErrorDetails());
      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
        action = Network::PostIoAction::Close;
      }
      break;
}

ENVOY_CONN_LOG(trace, "write returns: {}", callbacks_->connection(), result.return_value_);
bytes_written += result.return_value_;
```

Reply:
of course you would :-P